[lens] Show field details on click

wylieconlon opened the pull request:
This shows a popover on click based on the field type and time range.

### Strings

Strings are sampled

<img width="290" alt="Screenshot 2019-09-18 16 28 41" src="https://user-images.githubusercontent.com/666475/65183592-c8492400-da31-11e9-8635-66cf48c31d20.png">

### For dates

We don't sample dates, this is showing exact data. The date formatters are not using the standard field formatters:

<img width="621" alt="Screenshot 2019-09-18 16 28 54" src="https://user-images.githubusercontent.com/666475/65183612-d303b900-da31-11e9-8f6a-fe87686c471b.png">

### Numbers

Numbers have a toggle which decides whether to show a histogram or top values based on the distribution of data. When 90% of the data is in the top 10 values, that is shown first:

<img width="317" alt="Screenshot 2019-09-18 16 31 01" src="https://user-images.githubusercontent.com/666475/65183664-eca50080-da31-11e9-9a26-b247813765f5.png">

Otherwise it shows a distribution:

<img width="310" alt="Screenshot 2019-09-18 16 30 54" src="https://user-images.githubusercontent.com/666475/65183620-d8610380-da31-11e9-8914-c613570b5798.png">

### Booleans

<img width="332" alt="Screenshot 2019-09-17 13 28 26" src="https://user-images.githubusercontent.com/666475/65064912-5c8e8a80-d94f-11e9-802a-51cf27fc4606.png">

Commit:
[lens] Initial Commit (#35627)

* [visualization editor] Initial Commit

* [lens] Add more complete initial state

* [lens] Fix type issues

* [lens] Remove feature control

* [lens] Bring back feature control and add tests

* [lens] Update plugin structure and naming per comments

* replace any usage by safe casting

* [lens] Respond to review comments

* [lens] Remove unused EditorFrameState type

wylieconlon committed:
[lens] Initial state for IndexPatternDatasource (#36052)

* [lens] Add first tests to indexpattern data source

* Respond to review comments

* Fix type definitions

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
[lens] Editor frame initializes datasources and visualizations (#36060)

* [lens] Editor frame initializes datasources and visualizations

* Respond to review comments

* Fix build issues

* Fix state management issue

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

chrisdavies committed:
[lens][draft] Lens/drag drop (#36268)

Add basic drag / drop component to Lens

flash1293 committed:
remove local package (#36456)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[lens] Native renderer (#36165)

* Add nativerenderer component

* Use native renderer in app and editor frame

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] No explicit any (#36515)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Implement basic editor frame state handling (#36443)

wylieconlon committed:
[lens] Load index patterns and render in data panel (#36463)

* [lens] Editor frame initializes datasources and visualizations

* Respond to review comments

* Fix build issues

* remove local package

* [lens] Load index patterns into data source

* Redo types for Index Pattern Datasource

* Fix one more type

* Respond to review comments

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

chrisdavies committed:
[draft] Lens/line chart renderer (#36827)

Expression logic for the Lens xy chart.

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

Commit:
[lens] Index pattern data panel (initial) (#37015)

* [lens] Index pattern switcher

* Respond to review comments

flash1293 committed:
[Lens] Editor state 2 (#36513)

Commit:
[lens] Dimension panel that generates columns (#37117)

* [lens] Dimension panel that generates columns

* Update from review comments

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

Commit:
[lens] Generate esdocs queries from index pattern (#37361)

* [lens] Generate esdocs queries from index pattern

* Remove unused code

wylieconlon committed:
Merge branch 'origin/master' into feature/lens

wylieconlon committed:
Update yarn.lock from yarn kbn bootstrap

flash1293 committed:
Merge branch 'master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

chrisdavies committed:
[Lens] Add basic Lens xy chart suggestions (#37030)

Basic xy chart suggestions

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Expression rendering (#37648)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Expression handling (#37876)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

chrisdavies committed:
[Lens] Lens/xy config panel (#37391)

Basic xy chart configuration panel

flash1293 committed:
[Lens] Xy expression building (#37967)

flash1293 committed:
[Lens] Initialize visualization with datasource api (#38142)

Commit:
[lens] Dimension panel lets users select operations and fields individually (#37573)

* [lens] Dimension panel lets users select operations and fields individually

* Split files and add tests

* Fix dimension labeling and add clear button

* Support more aggregations, aggregation nesting, rollups, and clearing

* Fix esaggs expression

* Increase top-level test coverage of dimension panel

* Update from review comments

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
[Lens] Rename columns (#38278)

chrisdavies committed:
[Lens] Lens/index pattern drag drop (#37711)

* Basic xy chart suggestions

* Re-apply XY config panel after force merge

* Initial integration of lens drag and drop

* Tweak naming, remove irellevant comment

* Tweaks per Wylie's feedback

* Add xy chart internationalization
Tweak types per Joe's feedback

* Update xy chart i18n implementation

* Fix i18n id

* Add drop tests to the lens index pattern

* improve tests

Commit:
[lens] Only allow aggregated dimensions (#38820)

* [lens] Only allow aggregated dimensions

* [lens] Index pattern suggest on drop

* Fully remove value

* Revert "[lens] Index pattern suggest on drop"

This reverts commit 604c6ed68ca394441ddafa662bdfc5f421de300c.

* Fix type errors

Commit:
[lens] Suggest on drop (#38848)

* [lens] Index pattern suggest on drop

* Add test for suggestion without date field

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
fix merge

flash1293 committed:
[Lens] Parameter configurations and new dimension config flow (#38863)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
fix eslint failure

Commit:
[lens] Fix build by updating saved objects and i18n (#39391)

* [lens] Update location of saved objects code

* Update internatationalization

* Remove added file

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

Commit:
[lens] Fix arguments to esaggs using booleans (#39462)

Commit:
[lens] Datatable visualization plugin (#39390)

* [lens] Datatable visualization plugin

* Fix merge issues and add tests

* Update from review

* Fix file locations

Commit:
[lens] Use first suggestion when switching visualizations (#39377)

Commit:
[lens] Label each Y axis with its operation label (#39461)

* [lens] Label each Y axis with its operation label

* Remove comment

* Add link to chart issue

flash1293 committed:
[Lens] Suggestion preview rendering (#39576)

flash1293 committed:
[Lens] Popover configs (#39565)

flash1293 committed:
[Lens] Basic layouting (#39587)

flash1293 committed:
remove datasource public API in suggestions (#39772)

chrisdavies committed:
[Lens] Basic save / load (#39257)

Add basic routing, save, and load to Lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
[lens] Fix lint error

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
[lens] Use node scripts/eslint.js --fix to fix errors

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

Commit:
[lens] Include link to lens from Visualize (#40542)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

Commit:
[lens] Support stacking in xy visualization (#40546)

* [lens] Support stacking in xy visualization

* Use chart type switcher for stacked and horizontal xy charts

* Clean up remaining isStacked code

* Fix type error

chrisdavies committed:
[Lens] Add xy split series support (#39726)

* Add split series to lens xy chart

Commit:
[lens] Lens Filter Ratio (#40196)

* WIP filter ratio

* Fix test issues

* Pass dependencies through plugin like new platform

* Pass props into filter ratio popover editor

* Provide mocks to filter_ratio popover test

* Add another test

* Clean up to prepare for review

* Clean up unnecessary changes

* Respond to review comments

* Fix tests

flash1293 committed:
[Lens] Terms order direction (#39884)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
fix types

chrisdavies committed:
[Lens] Data panel styling and optimizations (#40787)

Style the data panel (mostly Joe Reuter's doing). Optimize a bunch of the Lens stack.

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Optimize dimension panel flow (#41114)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] re-introduce no-explicit-any (#41454)

flash1293 committed:
[Lens] No results marker (#41450)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

Commit:
[lens] Support layers for visualizing results of multiple queries (#41290)

* [lens] WIP add support for layers

* [lens] WIP switch to nested tables

* Get basic layering to work

* Load multiple tables and render in one chart

* Fix priority ordering

* Reduce quantity of linting errors

* Ensure that new xy layer state has a split column

* Make the "add" y / split the trailing accessor

* Various fixes for datasource public API and implementation

* Unify datasource deletion and accessor removal

* Fix broken scss

* Fix xy visualization TypeScript errors?

* Build basic suggestions

* Restore save/load and fix typescript bugs

* simplify init routine

* fix save tests

* fix persistence tests

* fix state management tests

* Ensure the data table is aligned to the top

* Add layer support to Lens datatable

* Give xy chart a default layer initially

* Allow deletion of layers in xy charts

* xy: Make split accessor singular
Remove commented code blocks

* Change expression type for lens_merge_tables

* Fix XY chart rendering expression

* Fix type errors relating to `layerId` in table suggestions

* Pass around tables for suggestions with associated layerIds

* fix tests in workspace panel

* fix editor_frame tests

* Fix xy tests, skip inapplicable tests
that will be implemented in a separate PR

* add some tests for multiple datasources and layers

* Suggest that split series comes before X axis in XY chart

* Get datatable suggestion working

* Adjust how xy axis labels are computed

* Datasource suggestions should handle layers and have tests

* Fix linting in XY chart and remove commented code

* Update snapshots from earlier change

* Fix linting errors

* More cleanup

* Remove commented code

* Test the multi-column editor

* XY Visualization does not need to track datasourceId

* Fix various comments

* Remove unused xy prop
Add datasource header to datatable config

* Use operation labels for XY chart

* Adding and removing layers is reflected in the datasource

* rewrote datasource state init

* clean up editor_frame frame api implementation

* clean up editor frame

flash1293 committed:
[Lens] Embeddable (#41361)

Commit:
[lens] Move XY chart config into popover and fix layering (#41927)

* [lens] Move XY chart config into popover and fix layering

* Fix tests

* Update style

* Change wrapping of layer settings popover

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Fix bugs in date_histogram and filter ratio (#42046)

flash1293 committed:
[Lens] Performance improvements (#41784)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
fix type error

flash1293 committed:
switch default size of terms operation to 3 (#42334)

flash1293 committed:
[lens] Improve suggestions for split series (#42052)

chrisdavies committed:
[lens] Add chart switcher (#42093)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
solve merge conflicts

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
fix test case

flash1293 committed:
[Lens] Allow only current visualization on field drop in workspace (#42344)

flash1293 committed:
[Lens] Remove indexpattern id on column (#42429)

Commit:
[lens] Implement app-level filtering and time picker (#42031)

* [lens] Implement app-level filtering and time picker

* More integration with filter bar

* Clean up test code and type errors

* Add frame level tests for syncing with app

* Add test coverage for app logic

* Simplify state management from app down

* Fix import errors

* Clarify whether properties are ids or titles for index pattern

* pass new saved object by ref

* add dirty state checking

* Fix tests

flash1293 committed:
[Lens] Add some tests around document handling in dimension panel (#42670)

flash1293 committed:
[Lens] Terms operation boolean support (#42817)

Commit:
[lens] Minor UX/UI improvements in Lens (#42852)

* Make dimension popover toggle when clicking button

* Without suggestions hide suggestion panel

Commit:
Add missing translations (#42921)

cchaos committed:
[Lens] Config panel design (#42980)

* Fix up design of config panel

Does not include config popover

Commit:
Remove a couple of non-null assertions (#43013)

* Remove a couple of non-null assertions

* Remove orphaned import

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
[Lens] Switch indexpattern manually (#42599)

cchaos committed:
[Lens] Update frame to put suggestions at the bottom (#42997)

flash1293 committed:
fix type errors

flash1293 committed:
switch indexpattern on layer if there is only a single empty one (#43079)

flash1293 committed:
[Lens] Suggest reduced versions of current data table (#42537)

flash1293 committed:
[Lens] Field formatter support (#38874)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

flash1293 committed:
Fix bugs

chrisdavies committed:
[Lens] Add bucket nesting editor to indexpattern (#42869)

chrisdavies committed:
[Lens] Remove unnecessary fields and indexing from mappings (#43285)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 committed:
[Lens] Xy scale type (#42142)

Commit:
[lens] Allow updater function to be used for updating state (#43373)

chrisdavies committed:
[Lens] Lens metric visualization (#39364)

Commit:
Fix axis rotation (#43792)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

Commit:
[Lens] Auto date histogram (#43775)

* Add auto date histogram

* Improve documentation and cleanup

* Add tests

* Change test name

flash1293 committed:
[Lens] Fix query bar integration (#43865)

chrisdavies committed:
Merge branch 'master' of github.com:elastic/kibana into feature/lens

flash1293 committed:
[Lens] Clean up operations code (#43784)

chrisdavies committed:
[Lens] Functional tests (#44279)

Foundational layer for lens functional tests. Lens-specific page objects are not in this PR.

chrisdavies committed:
[Lens] Add Lens visualizations to Visualize list (#43398)

flash1293 committed:
[Lens] Suggestion improvements (#43688)

wylieconlon committed:
[lens] Calculate existence of fields in datasource

wylieconlon committed:
[lens] Show field details on click

Commit:
[lens] Calculate existence of fields in datasource (#44422)

* [lens] Calculate existence of fields in datasource

* Fix route registration

* Add page object and use existence in functional test

* Simplify layout of filters for index pattern

* Respond to review feedback

* Update class names

* Use new URL constant

* Fix usage of base path

* Fix lint errors

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/850/)

wylieconlon committed:
Add basic API test and make trigger into draggable

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/888/)

flash1293 committed:
[Lens ] Preview metric (#43755)

flash1293 committed:
format filter ratio as percentage (#44625)

flash1293 committed:
[Lens] Remove datasource suggestion id (#44495)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

chrisdavies committed:
[Lens] Make breadcrumbs look and feel like Visualize (#44258)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Improve click target for popover

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1137/)

wylieconlon committed:
[lens] Fix breakage from app-arch movements (#44720)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
[lens] Fix type error in test from merge

wylieconlon committed:
Merge branch origin/feature/lens

Commit:
[lens] Fix registration of embeddable (#45171)

wylieconlon committed:
Upgrade EUI and fix typescript breakages

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2063/)

wylieconlon committed:
Undo EUI changes

chrisdavies committed:
[Lens] Functional tests (#44814)

Basic functional tests for Lens, by no means comprehensive. This is more of a smokescreen test of some normal use cases.

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

wylieconlon committed:
Fix snapshots

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

wylieconlon committed:
Fix issues with types and tests

wylieconlon committed:
Add back accidentally committed code

chrisdavies commented:
Nice work. I'm just starting the review, but I thought I'd note some UX feedback here.

It currently flies out from the word, but I think it should fly out from the field's container, so it's always aligned in the same way.

![image](https://user-images.githubusercontent.com/833377/64641060-5c374200-d3d9-11e9-9084-eaef6ae51165.png)

Maybe should instead look more like this:

![image](https://user-images.githubusercontent.com/833377/64641156-91dc2b00-d3d9-11e9-84f4-a721805a7b15.png)

And the field should probably be hi-lighted in some way.

I was also a bit confused at first, because I was looking at the last 15 minutes, so my flyouts looked a bit off:

![image](https://user-images.githubusercontent.com/833377/64641245-c0f29c80-d3d9-11e9-86a8-5d4fdfa6fef8.png)

I don't know what we should do about that, but that is a really weird flyout!

cchaos commented:
Looks like there's still some work to do on these. There seems to be an open ticket https://github.com/elastic/kibana/issues/44152

Commit:
[lens] Add Lens to CODEOWNERS (#45296)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2238/)

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -24,7 +87,21 @@ function wrapOnDot(str?: string) {
 }
 
-export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemProps) {
+export function FieldItem({
```
This component feels a bit too complex now, and I think it would benefit from being broken into smaller pieces. Those pieces can still all live in this file, but it's currently pretty hard to follow all of the indentations, brackets, etc.

If I were to make a list of the things / responsibilities here, it would look something like this:

```js
// Selectors
function wrapOnDot(str);
function getFormatter(opts);

// Actions / effects
function loadFieldStats(opts);

// Components
function FieldItem(props); // The outer component
function FieldStats(props); // The flyout container
function HilightedText(props); // The hilighted text (currently lines 105-119)
function FieldStatsSwitcher(props); // Maybe the dropdown / chooser
function FieldStatsHistogram(props); 
function FieldStatsTopValues(props);
```

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -95,6 +96,116 @@ export async function initStatsRoute(setup: CoreSetup) {
+        body: schema.object(
+          {
+            // query: schema.object({}),
+            earliest: schema.string(),
```
I find earliest / latest to be a bit confusing. Not sure what we should use, but maybe just keep the date range convention used elsewhere: `fromDate` and `toDate`. Also, should the schema check to ensure these are dates, or should we just let Elasticsearch handle that for us?

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -95,6 +96,116 @@ export async function initStatsRoute(setup: CoreSetup) {
     }
   );
+
+  router.post(
```
It's probably best to give each endpoint its own file, so the code is focused and contained, otherwise it starts to blur together when looking through the one file.

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -95,6 +96,116 @@ export async function initStatsRoute(setup: CoreSetup) {
+            ),
+          });
+        }
+        return res.ok();
```
I wonder if we should error if we're called with a field type we don't support?

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -95,6 +96,116 @@ export async function initStatsRoute(setup: CoreSetup) {
+          },
+        };
+
+        if (field.type === 'number') {
```
If I'm following this correctly, this if block could be simplified:

```js
  if (field.type !== 'number' && field.type !== 'date' && field.type !== 'string') {
    // Should this be an error?
    return res.ok();
  }

  const search = (aggs: unknown) => requestClient.callAsCurrentUser('search', {
    index: req.params.indexPatternTitle,
    body: {
      query: filters,
      aggs,
    },
    size: 0,
  });

  return res.ok({
    body: field.type === 'number' ? await getNumberHistogram(search, field)
      : field.type === 'string' ? await getStringSamples(search, field)
      : await getDateSamples(search, field, { earliest, latest }),
  });
```

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -164,3 +275,170 @@ export function recursiveFlatten(
   return returnTypes;
 }
+
+export async function getNumberHistogram(
```
The interleaving of code and type definitions interfered with my reading of these functions. I'd suggest breaking the type definitions out. Also, you can make the search function parameterized, which helps a bit:

```js
type MinMaxResult = AggregationSearchResponse<
  unknown,
  {
    body: {
      aggs: {
        sample: {
          sampler: {};
          aggs: {
            min_value: {
              min: { field: string };
            };
            max_value: {
              max: { field: string };
            };
          };
        };
      };
    };
  }
>;

type NumericHistogramResult = AggregationSearchResponse<
  unknown,
  {
    body: {
      aggs: {
        sample: {
          sampler: {};
          aggs: {
            histo: {
              histogram: { field: string; interval: number };
            };
          };
        };
      };
    };
  }
>;

export async function getNumberHistogram(
  aggSearchWithBody: <T>(body: unknown) => Promise<T>,
  field: { name: string; type: string; esTypes?: string[] }
) {
  const minMaxResult = await aggSearchWithBody<MinMaxResult>({
    sample: {
      sampler: { shard_size: 5000 },
      aggs: {
        min_value: {
          min: { field: field.name },
        },
        max_value: {
          max: { field: field.name },
        },
        top_values: {
          terms: { field: field.name, size: 10 },
        },
      },
    },
  });

  const minValue = minMaxResult.aggregations!.sample.min_value.value;
  const maxValue = minMaxResult.aggregations!.sample.max_value.value;
  const terms = minMaxResult.aggregations!.sample.top_values;

  let histogramInterval = (maxValue! - minValue!) / 10;

  if (Number.isInteger(minValue!) && Number.isInteger(maxValue!)) {
    histogramInterval = Math.ceil(histogramInterval);
  }

  if (histogramInterval === 0) {
    return {
      topValues: terms,
      histogram: { buckets: [], docCount: minMaxResult.aggregations!.sample.doc_count },
    };
  }

  const histogramResult = await aggSearchWithBody<NumericHistogramResult>({
    sample: {
      sampler: { shard_size: 5000 },
      aggs: {
        histo: {
          histogram: {
            field: field.name,
            interval: histogramInterval,
          },
        },
      },
    },
  });

  return {
    // Probably need to convert the histogram's props from snake_case to camelCase here...
    histogram: histogramResult.aggregations!.sample,
    topValues: terms,
  };
}
```

wylieconlon replied:
I've actually done even better than this, check it out!

flash1293 replied:
I have one suggestion to reduce the number of casts and unknowns by unifying it in one place: Instead of passing in a function to do the search, the individual getter functions take a search dependency object that they pass to a module scope search function that does the cast once:
```ts
// post handler

  // collect everything necessary to do searches 
  const searchDeps = { requestClient, index: req.params.indexPatternTitle, filters };

  if (field.type === 'number') {
    return res.ok({
      // pass it to the actual getter
      body: await getNumberHistogram(searchDeps, field),
  });

//...

// these are the dependencies necessary for searches
interface SearchDependencies {
  requestClient: ScopedClusterClient;
  index: string;
  filters: unknown;
}

// search function itself has a generic type that makes it possible to infer
// the return value from the aggs parameter
function search<T>({ requestClient, index, filters }: SearchDependencies, aggs: T) {
  return requestClient.callAsCurrentUser('search', {
    index,
    body: {
      query: filters,
      aggs,
    },
    // The hits total changed in 7.0 from number to object, unless this flag is set
    // this is a workaround for elasticsearch response types that are from 6.x
    restTotalHitsAsInt: true,
    size: 0,
   // still have to cast here because callAsCurrentUser won't do it for us
  }) as Promise<AggregationSearchResponse<unknown, { body: { aggs: T } }>>;
}

// ...

export async function getNumberHistogram(
  searchDeps: SearchDependencies,
  field: { name: string; type: string; esTypes?: string[] }
): Promise<FieldStatsResponse> {
  // ...
  // can use search here without any casts, everything is inferred from searchBody
  const minMaxResult = await search(searchDeps, searchBody);
  // ...
}
```

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -164,3 +275,170 @@ export function recursiveFlatten(
+    }
+  >;
+
+  const minValue = minMaxResult.aggregations!.sample.min_value.value;
```
Just double checking: is it possible for aggregations to ever be null here?

wylieconlon replied:
It should always return a body if the aggregation query ran at all, but maybe I'm missing an error handler somewhere around the request

dgieselaar replied:
`aggregations` will be undefined if no matching indices were found. It's an edge case, but it happens, so it's reflected in the type.

chrisdavies commented on x-pack/legacy/plugins/lens/server/routes/index_stats.ts:
```diff
@@ -164,3 +275,170 @@ export function recursiveFlatten(
+
+  const histogramResult = (await aggSearchWithBody({
+    sample: {
+      sampler: { shard_size: 5000 },
```
Maybe SHARD_SIZE should be a const, since it's currently a magic number that exists in several places in this file.

chrisdavies reviewed (commented):
I think this needs a bit more work.

The logic is all reasonable. The UX has a bit of styling still to be done, but so does the rest of the app.

The main problem I have is that the code itself was pretty hard to grok due to length and a lot of nesting, and in some cases, complex type definitions obscuring the flow of logic. I'd definitely extract any complex type definitions out of functions and into the module(s).

wylieconlon commented on x-pack/legacy/plugins/apm/typings/elasticsearch.ts:
```diff
@@ -71,6 +72,10 @@ declare module 'elasticsearch' {
     >;
   };
 
+  type SamplerAggregation<SubAggregationMap> = SubAggregation<
```
It's equivalent to `type SamplerAggregation<T> = SubAggregation<T>;` - the extra name is not necessary but makes this line easier later, instead of `{ sampler: SubAggregation<AggregationOption[AggregationName]>; }`

dgieselaar replied:
These types are a bit raw, it was our first attempt of getting aggregation inference. I'm actually super impressed that you've managed to make it work 😅

wylieconlon commented on x-pack/legacy/plugins/apm/typings/elasticsearch.ts:
```diff
@@ -128,6 +133,7 @@ declare module 'elasticsearch' {
         cardinality: {
           value: number;
         };
+        sampler: SamplerAggregation<AggregationOption[AggregationName]>;
```
@chrisdavies This is where it's used- I think the alias is helpful here

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -4,17 +4,80 @@
+    key: number;
+    doc_count: number;
+  }>;
+  sum_other_doc_count?: number;
```
Would you say something different if I were using pre-defined types from elasticsearch? But you're right that I could do this.

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -41,6 +118,82 @@ export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemP
+    formatter = fieldFormats.getDefaultInstance(field.type, field.esTypes);
+  }
+
+  useEffect(() => {
```
I like this way of looking at it, will update

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -41,6 +118,82 @@ export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemP
+        }
+      })
+      .catch(() => {
+        setState(s => ({ ...s, isLoading: false }));
```
Probably should...

wylieconlon committed:
Change popover styles and simplify types on server

wylieconlon commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -194,31 +186,44 @@ export function FieldItem({
+    <EuiPopover
+      id="lnsFieldListPanel__field"
+      // Wait until @elastic/eui is at least version 13.7.0 to use display="block"
+      display="block"
```
We really do need display="block" here, so I'm hoping to get #45052 merged into master first

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2285/)

flash1293 committed:
Merge remote-tracking branch 'upstream/master' into feature/lens

wylieconlon committed:
[lens] Fix visualization alias registration

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

Commit:
[lens] Fix usage of EUI after typescript upgrade (#45404)

* [lens] Fix usage of EUI after typescript upgrade

* Use local fix instead of workaround

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

wylieconlon committed:
Process results more

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2507/)

chrisdavies reviewed (approved):
LGTM. Just one place where I think you left in some commented-out code.

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Fix API errors

wylieconlon committed:
Fix API tests and use sampleCount

dgieselaar commented on x-pack/legacy/plugins/apm/typings/elasticsearch.ts:
```diff
@@ -82,7 +88,11 @@ declare module 'elasticsearch' {
+        value_count: AggregatedValue;
+        terms: BucketAggregation<
+          AggregationOption[AggregationName],
+          string | number
```
this is correct (TIL!) but it's causing some type errors downstream. Adding a `.toString()` should be enough in most of the cases.

dgieselaar reviewed (dismissed):
Happy to re-review & approve once the type issues are fixed. (I'm off until Tuesday, so if you need any assistance before that, feel free to ping others from APM UI).

flash1293 commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -41,22 +97,286 @@ export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemP
+            <Axis
+              id={getAxisId('key')}
+              position={Position.Left}
+              tickFormat={d => formatter.convert(d)}
```
**Optional**: The default date format is suboptimal for most cases - we could use
```ts
{
  niceTimeFormatter,
} from '@elastic/charts';
```

in this case

flash1293 commented on x-pack/legacy/plugins/lens/server/routes/field_stats.ts:
```diff
@@ -0,0 +1,259 @@
+  const minValue = minMaxResult.aggregations!.sample.min_value.value;
+  const maxValue = minMaxResult.aggregations!.sample.max_value.value;
+  const terms = minMaxResult.aggregations!.sample.top_values;
+  const topValuesBuckets = bucketsToResponse<{ terms: { field: string; size: number } }>(terms);
```
nit: The type parameter should get inferred, no need to spell it out

flash1293 commented on x-pack/legacy/plugins/lens/server/routes/field_stats.ts:
```diff
@@ -0,0 +1,259 @@
+      },
+    },
+  };
+  const histogramResult = (await aggSearchWithBody(histogramBody)) as AggregationSearchResponse<
```
```
AggregationSearchResponse<
    unknown,
    { body: { aggs: typeof histogramBody } }
  >
```

is probably long enough for a type alias

flash1293 commented on x-pack/legacy/plugins/lens/server/routes/field_stats.ts:
```diff
@@ -0,0 +1,259 @@
+
+  return {
+    count: topValuesResult.hits.total,
+    sampleCount: topValuesResult.aggregations!.sample.sample_count.value!,
```
The problem with this is that array fields are counted with the number of values, not just 1. This can lead to more than 100% in the UI because it's compared to the total number of hits:
<img width="630" alt="Screenshot 2019-09-13 at 11 04 25" src="https://user-images.githubusercontent.com/1508364/64852678-1b7a3b80-d61a-11e9-8d3e-02b899335e99.png">

Not sure about an easy way out of this, maybe we need another aggregation for that to explicitly check for empty fields?

flash1293 reviewed (commented):
Awesome work, @wylieconlon - this is looking great! Had some smaller remarks, the only semi-major one is the problem with sample count

flash1293 commented:
@wylieconlon About types in the server: I have one suggestion to reduce the number of casts and unknowns by unifying it in one place: Instead of passing in a function to do the search, the individual getter functions take a search dependency object that they pass to a module scope search function that does the cast once:
```ts
// post handler

  // collect everything necessary to do searches 
  const searchDeps = { requestClient, index: req.params.indexPatternTitle, filters };

  if (field.type === 'number') {
    return res.ok({
      // pass it to the actual getter
      body: await getNumberHistogram(searchDeps, field),
  });

//...

// these are the dependencies necessary for searches
interface SearchDependencies {
  requestClient: ScopedClusterClient;
  index: string;
  filters: unknown;
}

// search function itself has a generic type that makes it possible to infer
// the return value from the aggs parameter
function search<T>({ requestClient, index, filters }: SearchDependencies, aggs: T) {
  return requestClient.callAsCurrentUser('search', {
    index,
    body: {
      query: filters,
      aggs,
    },
    // The hits total changed in 7.0 from number to object, unless this flag is set
    // this is a workaround for elasticsearch response types that are from 6.x
    restTotalHitsAsInt: true,
    size: 0,
   // still have to cast here because callAsCurrentUser won't do it for us
  }) as Promise<AggregationSearchResponse<unknown, { body: { aggs: T } }>>;
}

// ...

export async function getNumberHistogram(
  searchDeps: SearchDependencies,
  field: { name: string; type: string; esTypes?: string[] }
): Promise<FieldStatsResponse> {
  // ...
  // can use search here without any casts, everything is inferred from searchBody
  const minMaxResult = await search(searchDeps, searchBody);
  // ...
}
```

wylieconlon committed:
Simplify document count calculation

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2905/)

Commit:
[lens] Fix usage of expressions plugin (#45544)

* [lens] Fix usage of expressions plugin

* Use updated exports from #45538

* Fix imports and mocha tests

* Use relative instead of absolute path to fix tests

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Fix API test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2948/)

wylieconlon committed:
Fix type errors and pass through new platform from top

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2978/)

Commit:
[lens] More cleanup from QueryBar changes in master (#45687)

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -41,22 +98,301 @@ export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemP
+  const otherCount =
+    state.sampledValues && bucketedValuesCount ? state.sampledValues - bucketedValuesCount : 0;
+
+  const euiButtonColor =
```
These two values are identical. Is this incidental or inherent? If the latter, we should consolidate into one value.

wylieconlon replied:
They're unfortunately not identical, buttons don't have a 'default' style and text doesn't have 'primary'

chrisdavies commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/field_item.tsx:
```diff
@@ -41,22 +98,301 @@ export function FieldItem({ field, indexPattern, highlight, exists }: FieldItemP
+                }
+              }}
+              onKeyPress={event => {
+                if (exists && event.key === 'ENTER') {
```
I think it would be better to have a helper function that opens the flyout and loads the data, then call that anywhere you're trying to toggle the flyout:

```js
const toggleFlyout = () => {
  if (!exists) {
    return;
  }
  
  setOpen(!infoIsOpen);
  if (!infoIsOpen) {
    fetchData();
   }
};
```

That reduces the likelihood of failing to fetch data when attempting to open the flyout.

chrisdavies reviewed (approved):
LGTM w/ possible tweaks.

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

wylieconlon committed:
Fix tests

wylieconlon committed:
Fix i18n issue

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3181/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3194/)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into feature/lens

flash1293 commented:
This looks mostly good to me. One thing I noticed is that it shows an empty overlay for boolean values:
<img width="768" alt="Screenshot 2019-09-17 at 14 06 44" src="https://user-images.githubusercontent.com/1508364/65040349-72eb1500-d954-11e9-82b3-97d07a6b1eda.png">

Commit:
[lens] Fix build and use new platform from entry points (#45834)

* [lens] Fix build and use new platform from entry points

* Fix params for existence route

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

elasticmachine committed:
Merge branch 'master' into feature/lens

wylieconlon committed:
Fix tests and add boolean/empty field support

wylieconlon committed:
Merge remote-tracking branch 'origin/feature/lens' into lens/field-popover

wylieconlon commented:
@flash1293 I fixed the boolean fields, thanks for testing that out.

@chrisdavies I actually find that it's important to allow the popover for fields that we are considering to be "empty" because they might be empty in the default time range, but not empty after the time range is changed.

sorenlouv reviewed (approved):
APM change looks good 👍

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3450/)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/field-popover

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3485/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3484/)

Commit:
Updated top values

Commit:
Fixed up overall layout including adding a popover header

Commit:
Fix up time display

cchaos commented:
Design PR4U:  https://github.com/wylieconlon/kibana/pull/6

I have also run into bugs, things I need input on, and other thoughts that I'll need to aggregate, but overall nice implementation. I'll add my other comments tomorrow as I've run out of time today

cchaos commented:
Here are some things that I ran into while working on the design.

## Bugs - Major

### Fields panel would stop loading fields randomly

I was doing a lot of clicking on different fields, sometimes switching index patterns, sometimes changing dates and refreshing the page. At certain points the fields panel would just show the "No fields" message no matter the index patter or date selection. There are also no console errors or any other error visible. It just seems to get borked.

<img src="https://d.pr/free/i/KueWLc+" />

### Date field doesn't seem to update for "Last 1 year" quick selection

<img src="https://d.pr/free/i/6rPluo+"  />

### Fields panel completely borks with some error

I'm not sure what I did to cause this error but maybe this screenshot will give some detail. You can also see that the fields panel stops rendering completely.

<img width="1340" alt="Screen Shot 2019-09-17 at 19 16 17 PM" src="https://user-images.githubusercontent.com/549577/65150128-6ff23200-d9f1-11e9-8137-201d3407a85a.png">


## Formatting "bugs"

### "100%" shouldn't have any decimal places

In fact since these are just really quick summaries, I would suggest rounding this percentage always to a whole number.

<img width="413" alt="Screen Shot 2019-09-17 at 18 30 40 PM" src="https://user-images.githubusercontent.com/549577/65149908-fc502500-d9f0-11e9-9307-6834443e042d.png">

### Consider rounding the values in the distribution graph axes

It makes it really hard to truly follow the values because you have to look at whether it's a dot or comma. I'd suggest that these values should always be whole numbers as well.

**Depending on the data of course.** If there are no values larger than 1, or the spread is very small then showing decimals is helpful. But if, like in the screenshot below, the spread is in the thousands, then the decimals are not helpful.

<img width="343" alt="Screen Shot 2019-09-17 at 19 03 17 PM" src="https://user-images.githubusercontent.com/549577/65150033-42a58400-d9f1-11e9-9851-eed27e10e8f0.png">

## Empty/missing value state questions

### Empty values in top values

I feel like we should put something in place of empty values instead of just leaving that area blank.

<img width="628" alt="Screen Shot 2019-09-17 at 18 36 12 PM" src="https://user-images.githubusercontent.com/549577/65150232-b21b7380-d9f1-11e9-8ef7-93d4b14b1d59.png">


### No top values

I think that's what this state is indicating? But I'm really not sure what the data should look like but I feel we should account for it. I've also seen this state with more than 1 document.

<img width="608" alt="Screen Shot 2019-09-17 at 19 13 53 PM" src="https://user-images.githubusercontent.com/549577/65150304-e0994e80-d9f1-11e9-8471-9be1584311fb.png">


## Text question

### @AlonaNadler I added titles to these popovers and changed "Histogram" to "Distribution" everywhere. But I couldn't think of a good title for the date field.

<img width="526" alt="Screen Shot 2019-09-17 at 19 15 22 PM" src="https://user-images.githubusercontent.com/549577/65150387-09214880-d9f2-11e9-82f5-1b65a70824bd.png">

Commit:
Merge pull request #6 from cchaos/lens/field-popover/design

Design PR

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3648/)

elasticmachine commented:
Pinging @elastic/kibana-app

wylieconlon commented:
@cchaos 

> Fields panel would stop loading fields randomly

Chris is working on this and other issues with the list of fields

> Date field doesn't seem to update for "Last 1 year" quick selection

This is an issue with the rendering, not with the data fetching. We need to render a chart that matches the date range the user selected. I should have this fixed in the next push.

> Fields panel completely borks with some error

Definitely an issue when the server returns no data. I should have this fixed.

> Formatting "bugs"

> "100%" shouldn't have any decimal places

Sure, I've removed the decimal places from all of these

> Consider rounding the values in the distribution graph axes

I see where you're coming from, but I am not sure I can realistically make changes to the formatting- these are the same formatters as Visualize uses. I've already done some work to make the intervals into "nicer" intervals, but this might need another pass.

> Empty/missing value state questions

> Empty values in top values

I think this is the same issue as above. I'm not sure how to present something that is consistent with the rest of Kibana without seeing cases like the one you're seeing.

> No top values

This looks like a bug. I think I have a fix for it.

wylieconlon committed:
Improve popover edge cases and respond to feedback

wylieconlon committed:
Fix field formatters for histogram

wylieconlon commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into lens/field-popover

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3740/)

wylieconlon committed:
Merge remote-tracking branch 'origin/master' into lens/field-popover

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3778/)

cchaos commented:
> Empty values in top values

> I think this is the same issue as above. I'm not sure how to present something that is consistent with the rest of Kibana without seeing cases like the one you're seeing.

Could we do something like this?

<img width="302" alt="Screen Shot 2019-09-19 at 13 18 43 PM" src="https://user-images.githubusercontent.com/549577/65265950-13217500-dae0-11e9-8c37-969a0cfc376e.png">

wylieconlon commented:
@cchaos Yes, I can do that. Are there any other changes that need to be done with this PR?

wylieconlon committed:
Use custom style for empty top values

cchaos reviewed (approved):
The rest looks good to me, thank you!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3936/)
